### PR TITLE
feat: add basic observable implementations of procedure interfaces

### LINKED ISSUE #6514: Create basic concrete implementations of the procedure data models
### Problem Statement

We need some basic implementations of the ProcedureModels so then we can start adding the other functional bits (i.e. triggering `doProcedureUpdate` and firing events.

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [X] I ran `npm run format` and `npm run lint`

### Resolves

Fixes #6514 

Dependent on #6488 

### Proposed Changes

Adds some basic implementations of the procedure interfaces. They don't do any of the important stuff yet (i.e. triggering `doProcedureUpdate` or firing events).

### Reason for Changes

Interfaces need to be implemented.

### Test Coverage

N/A

### Documentation

N/A

### Additional Information

N/A

### CODE REVIEW

**Review by cpcallen — COMMENTED** [review]
A preliminary review pending discussion about #6488.

File: core/procedures/observable_procedure_model.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    return this;
+  }
+
+  /**
+   * Sets the return type(s) of the procedure.
+   *
+   * Pass null to represent a procedure that does not return.
+   */
+  setReturnTypes(types: string[]|null): this {
+    this.returnType = types;
```
Comment (maribethb) [inline core/procedures/observable_procedure_model.ts:52]: please make them both singular or both plural as discussed offline :)

**Review by maribethb — APPROVED** [review]

File: core/procedures/observable_procedure_model.ts (hunk 1 above)
Comment (BeksOmega) [reply core/procedures/observable_procedure_model.ts]: Done!

**Review by cpcallen — COMMENTED** [review]
I found this tab spinning, with my approval from yesterday apparently unsubmitted.  Grr...

**Comment by cpcallen** [comment]
> I found this tab spinning, with my approval-with-comments from yesterday apparently unsubmitted. Grr...

File: core/procedures/observable_parameter_model.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+export class ObservableParameterModel implements IParameterModel {
+  private id: string;
+  private variable: VariableModel;
+
+  constructor(
+      private readonly workspace: Workspace, name: string, id?: string) {
+    this.id = id ?? genUid();
+    this.variable = workspace.createVariable(name);
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/procedures/observable_parameter_model.ts:20]: I will look into this.

File: core/procedures/observable_parameter_model.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * Unimplemented. The built-in ParameterModel does not support typing.
+   * If you want your procedure blocks to have typed parameters, you need to
+   * implement your own ParameterModel.
+   */
+  setTypes(_types: string[]): this {
+    console.warn(
+        'The built-in ParameterModel does not support typing. You need to ' +
+        'implement your own custom ParameterModel.');
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/procedures/observable_parameter_model.ts:41]: Error sgtm!

File: core/procedures/observable_procedure_map.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  /**
+   * Adds the given ProcedureModel to the map of procedure models, so that
+   * blocks can find it.
+   */
+  add(proc: IProcedureModel): this {
+    // TODO(#6156): Fire events.
+    return this.set(proc.getId(), proc);
+  }
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/procedures/observable_procedure_map.ts:49]: Hmm true, I think I'll leave this for now and add a TODO linked to... #6526 to reconsider once I actually start using this interface.

**Pull request merged**